### ISSUE TITLE
ValueError: Sign not allowed in string format specifier

### ISSUE THREAD

**Issue by KoenVanDerHauw** [pastas/pastas#427]
For a certain time series (nothing special) pastas gives the following error in model.py:
  File "C:\ProgramData\Anaconda3\envs\imod-pastas\lib\site-packages\pastas\model.py", line 740, in solve
    print(self.fit_report(output=output))
  File "C:\ProgramData\Anaconda3\envs\imod-pastas\lib\site-packages\pastas\model.py", line 1563, in fit_report
    f"{val1:<8}{val2:<23}{val3:<9}"
**ValueError: Sign not allowed in string format specifier**

It seems that a calculated column width gets negative for this time series which results in an error in 'f"{val4:>{wspace+len_val4}}\n"' at line 1564 in model.py. It can be fixed for example by ensuring that wspace is positive and adding the following, just before the for-loop in line 1561:
        if (wspace < 0):
            wspace = 0

Python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:30:19) [MSC v.1929 64 bit (AMD64)]
Numpy version: 1.22.3
Scipy version: 1.9.1
Pandas version: 1.4.3
Pastas version: 0.20.0
Matplotlib version: 3.5.3

**Comment by raoulcollenteur** [comment]
Hi Koen,

Thanks for reporting @KoenVanDerHauw ! Can you provide an example of where this happens, perhaps with some data and a short script? Curious to see when exactly this happens, perhaps we can try and catch the error earlier in the process. Thanks!

https://github.com/pastas/pastas/blob/1cd9fd94502920de8ae1bb03a902fd0d5ab25354/pastas/model.py#L1561

**Comment by KoenVanDerHauw** [comment]
Hi Raoul,

A bit stupid, but for some reason the error doesn’t occur anymore with the same series and script that gave an error yesterday all the time. Maybe something changed in the meteo series in the meantime, that is the only external input. We retrieve the meteo data via “from_nearest_xy()”. A bad model was found, by the way, I think the meteo series was too short. Possibly this resulted in exceptional model statistics that gave an unexpected column width in the code.

It seemed that the variable ‘width’ in model.py became too small which resulted in ‘wspace+len_val4 < 0’. I think I saw that some of the parameter values was empty in the line ‘width=len(‘parameters.__str__().split("\n")[1])’.

I’ll send you the series (an IPF-file with a separate timeseries file) and below some model statistics, maybe you notice something weird (although these are from this morning, yesterday they may have been somewhat different).

Fit report GWL                                    Fit Statistics
================================================================
nfev    95                     EVP                          2.66
nobs    1189                   R2                          -0.00
noise   True                   RMSE                         0.39
tmin    2012-06-23 00:00:00    AIC                      -8380.54
tmax    2016-12-31 00:00:00    BIC                      -8350.05
freq    D                      Obj                          0.51
warmup  3650 days 00:00:00     ___
solver  LeastSquares           Interp.                        No

Parameters (6 optimized)
================================================================
                  optimal               stderr     initial  vary
rainevap_A   1.735111e+04          ±361729.61%  219.568466  True
rainevap_n   9.999859e+01            ±3403.44%    1.000000  True
rainevap_a   6.682850e+01            ±4484.90%   10.000000  True
rainevap_f  -6.597546e-09  ±52650030191532.16%   -1.000000  True
constant_d   1.829534e+01               ±1.24%   18.381615  True
noise_alpha  2.707617e+02              ±61.43%    1.000000  True

Warnings! (2)
================================================================
Parameter 'rainevap_f' on upper bound: 0.00e+00
Response tmax for 'rainevap' > than calibration period

Van: Raoul Collenteur ***@***.***>
Verzonden: donderdag 1 september 2022 07:03
Aan: pastas/pastas ***@***.***>
CC: Hauw, Koen van der ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [pastas/pastas] ValueError: Sign not allowed in string format specifier (Issue #427)


Hi Koen,

Thanks for reporting @KoenVanDerHauw<https://urldefense.com/v3/__https:/github.com/KoenVanDerHauw__;!!HBVxBjZwpQ!zSwW2A7Q6ZJLblGT8YBy-KG9JQAmY4RdaCoCMcVDcXvvxi6uvRW0zZpcZa3fvBuof1BzVBOM0t0o36z5573O0V3xYA3crEQ$> ! Can you provide an example of where this happens, perhaps with some data and a short script? Curious to see when exactly this happens, perhaps we can try and catch the error earlier in the process. Thanks!

https://github.com/pastas/pastas/blob/1cd9fd94502920de8ae1bb03a902fd0d5ab25354/pastas/model.py#L1561<https://urldefense.com/v3/__https:/github.com/pastas/pastas/blob/1cd9fd94502920de8ae1bb03a902fd0d5ab25354/pastas/model.py*L1561__;Iw!!HBVxBjZwpQ!zSwW2A7Q6ZJLblGT8YBy-KG9JQAmY4RdaCoCMcVDcXvvxi6uvRW0zZpcZa3fvBuof1BzVBOM0t0o36z5573O0V3x0SgyMto$>

**Comment by dbrakenhoff** [comment]
A while back we were getting the same error while running pytest. Then it turned out to be that one of the methods we used to calculate the width of the fit report was dependent on the width of the console it was being run in. So we didn't get any errors when running it ourselves in some kind of console or whatever, but somehow pytest was causing issues. Not sure if that is relevant here, because I had hoped to fix that, but I recognize the error message. 

I think the line you highlight Raoul is the only one where it could theoretically go wrong...(?), as the parameters DataFrame determines the width of the fit report, so that should always be okay. So one of the numbers in the fit dictionary might be the culprit?

Perhaps this is also an opportunity to deal with those large std errors that sometimes make the report ugly by writing our own to_string() converter that converts large numbers to scientific notation, while keeping the rest as is.

**Comment by KoenVanDerHauw** [comment]
Hallo David,

I do use a separate monitor today, which sometimes can a different console width, but even without it, I don’t get the error anymore. Yesterday I got the error both when running the script from a batchfile and within Spyder.

I think David is right and one of the numbers in the fit dictionary caused the error.

I isolated that time series when I got the error, but I’m testing pastas and updating our script for another 10.000 times series. When I get the error again, I’ll let you know.

Met vriendelijke groet,
Koen van der Hauw
Adviseur Geohydrologie
Sweco Nederland B.V. | Arnhem
M +31 6 53696100
***@***.***

Van: Davíd Brakenhoff ***@***.***>
Verzonden: donderdag 1 september 2022 09:38
Aan: pastas/pastas ***@***.***>
CC: Hauw, Koen van der ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [pastas/pastas] ValueError: Sign not allowed in string format specifier (Issue #427)


A while back we were getting the same error while running pytest. Then it turned out to be that one of the methods we used to calculate the width of the fit report was dependent on the width of the console it was being run in. So we didn't get any errors when running it ourselves in some kind of console or whatever, but somehow pytest was causing issues. Not sure if that is relevant here, because I had hoped to fix that, but I recognize the error message.

I think the line you highlight Raoul is the only one where it could theoretically go wrong...(?), as the parameters DataFrame determines the width of the fit report, so that should always be okay. So one of the numbers in the fit dictionary might be the culprit?

Perhaps this is also an opportunity to deal with those large std errors that sometimes make the report ugly by writing our own to_string() converter that converts large numbers to scientific notation, while keeping the rest as is.
